File: blind-75/longest_repeating_char_replacement.py
# leetcode 424

from typing import List
from collections import defaultdict

class Solution:
	def characterReplacement(self, s: str, k: int) -> int:
		charFrequency = defaultdict(int)
		left, right, longestLength = 0, 0, 0
		mostFreqeunt = []

		while(right < len(s)):
			charFrequency[s[right]] += 1

			if (len(mostFreqeunt) == 0):
				mostFreqeunt.append(s[right])
				mostFreqeunt.append(1)
			else:
				if (charFrequency.get(s[right]) >= mostFreqeunt[1]):
					mostFreqeunt[0] = s[right]
					mostFreqeunt[1] = charFrequency.get(s[right])

			# print(f"current charFrequency {charFrequency}")
			currWinLen = (right - left) + 1
			if (currWinLen - mostFreqeunt[1]) <= k:
				longestLength = max(longestLength, currWinLen)
			else:
				while True:
					maxLen = 0
					charFrequency[s[left]] -= 1

					# update the mostFrequent data
					for key, value in charFrequency.items():
						print(f"{key}, {value}")
						if value >= maxLen:
							mostFreqeunt[0] = key
							mostFreqeunt[1] = value
							maxLen = value

					left += 1
					currWinLen = (right - left) + 1
					# print(f"currWinLen {currWinLen} mostFreqeunt[1] {mostFreqeunt[1]}")
					if (currWinLen - mostFreqeunt[1]) <= k: break

			right += 1

		return longestLength


obj = Solution()

print(obj.characterReplacement("ABAB", 2)) # should output 4
print(obj.characterReplacement("AABABBA", 1)) # should output 4

File: blind-75/longest_repeating_char_replacement.java
// leetcode 424

import java.util.*;

class Solution {
	class MostFrequent {
		public char ch;
		public int frequency;

		MostFrequent() {}
		MostFrequent(char ch, int frequency) {
			this.ch = ch;
			this.frequency = frequency;
		}

		public String toString() {
			return "(" + ch + " " + frequency + ")";
		}
	}

	public int characterReplacement(String s, int k) {
		int longestLength = 0;
		int left = 0, right = 0;
		Map<Character, Integer> charFrequency = new HashMap<>();
		MostFrequent mostFreq = null;

		while(right < s.length()) {
			char ch = s.charAt(right);
			charFrequency.put(ch, charFrequency.getOrDefault(ch, 0) + 1);
			
			if (mostFreq == null) {
				mostFreq = new MostFrequent(ch, 1);
			} else {
				int currCharFreq = charFrequency.getOrDefault(ch, 0);
				if (mostFreq.frequency <= currCharFreq) {
					mostFreq.ch = ch;
					mostFreq.frequency = currCharFreq;
				}
			}
			// System.out.println("left : " + left + " right : " + right + " mostFreq " + mostFreq);

			int currWinLen = (right - left) + 1;
			if ((currWinLen - mostFreq.frequency) <= k) {
				longestLength = Math.max(longestLength, currWinLen);
			} else {
				do {
					int maxFreq = 0;
					char leftChar = s.charAt(left);
					charFrequency.put(leftChar, charFrequency.get(leftChar) - 1);

					// after the above change most freq can be changed so update it again
					for (Map.Entry<Character, Integer> entry : charFrequency.entrySet()) {
						if (entry.getValue() >= maxFreq) {
							mostFreq.ch = entry.getKey();
							mostFreq.frequency = entry.getValue();
							maxFreq = entry.getValue();
						}
					}
					left++;
					currWinLen = (right - left) + 1;
					// System.out.println("current charFrequency " + charFrequency);
					// System.out.println("window was invalid, shrinking left and updating mostFreq " + " left : " + left + " right " + right + " mostFreq " + mostFreq);
				} while ((currWinLen - mostFreq.frequency) > k);
			}

			right++;
		}

		return longestLength;
    }
}


class Main {
	public static void main(String[] args) {
		Solution obj = new Solution();

		System.out.println(obj.characterReplacement(new String("ABAB"), 2));
		System.out.println(obj.characterReplacement(new String("AABABBA"), 1));
	}
}